core/src/main/java/org/infinispan/interceptors/CacheWriterInterceptor.java

```diff
@@ -201,6 +208,75 @@ public Object visitPutMapCommand(InvocationContext ctx, PutMapCommand command) t
       return returnValue;
    }
 
+   @Override
+   public Object visitReadWriteKeyCommand(InvocationContext ctx, ReadWriteKeyCommand command) throws Throwable {
+      return visitWriteCommand(ctx, command);
+   }
+
+   @Override
+   public Object visitReadWriteKeyValueCommand(InvocationContext ctx, ReadWriteKeyValueCommand command) throws Throwable {
+      return visitWriteCommand(ctx, command);
+   }
+
+   @Override
+   public Object visitWriteOnlyKeyCommand(InvocationContext ctx, WriteOnlyKeyCommand command) throws Throwable {
+      return visitWriteCommand(ctx, command);
+   }
+
+   private <T extends DataWriteCommand & ParamsCommand> Object visitWriteCommand(InvocationContext ctx, T command) throws Throwable {
+      Object retval = invokeNextInterceptor(ctx, command);
+      if (!isStoreEnabled(command) || ctx.isInTxScope() || !command.isSuccessful()) return retval;
+      if (!isProperWriter(ctx, command, command.getKey())) return retval;
+
+      Param<PersistenceMode> persistMode = command.getParams().get(PersistenceMode.ID);
+      switch (persistMode.get()) {
+         case PERSIST:
+            Object key = command.getKey();
+            CacheEntry entry = ctx.lookupEntry(key);
+            if (entry != null) {
+               if (entry.isRemoved()) {
+                  boolean resp = persistenceManager.deleteFromAllStores(key, BOTH);
+                  if (getLog().isTraceEnabled()) getLog().tracef("Removed entry under key %s and got response %s from CacheStore", key, resp);
+               } else if (entry.isChanged()) {
+                  storeEntry(ctx, key, command);
+               }
+            }
+         case SKIP:
+            log.trace("Skipping cache store since persistence mode parameter is SKIP");
+      }
+      return retval;
+   }
+
+   @Override
+   public Object visitWriteOnlyManyEntriesCommand(InvocationContext ctx, WriteOnlyManyEntriesCommand command) throws Throwable {
```

Comment:
I didn't have time to complete that in time for 8.0, already noted as todo for 8.1 in [ISPN-5704](https://issues.jboss.org/browse/ISPN-5704).